apollo-router/src/query_planner/bridge_query_planner.rs

```diff
@@ -895,70 +890,48 @@ impl BridgeQueryPlanner {
             });
         }
 
-        {
-            let operation = doc
-                .executable
-                .operations
-                .get(key.operation_name.as_deref())
-                .ok();
-            let mut has_root_typename = false;
-            let mut has_schema_introspection = false;
-            let mut has_other_root_fields = false;
-            if let Some(operation) = operation {
-                for field in operation.root_fields(&doc.executable) {
-                    match field.name.as_str() {
-                        "__typename" => has_root_typename = true,
-                        "__schema" | "__type" if operation.is_query() => {
-                            has_schema_introspection = true
-                        }
-                        _ => has_other_root_fields = true,
-                    }
-                }
-                if has_root_typename && !has_schema_introspection && !has_other_root_fields {
-                    // Fast path for __typename alone
-                    if operation
-                        .selection_set
-                        .selections
-                        .iter()
-                        .all(|sel| sel.as_field().is_some_and(|f| f.name == "__typename"))
-                    {
-                        let root_type_name: serde_json_bytes::ByteString =
-                            operation.object_type().as_str().into();
-                        let data = Value::Object(
-                            operation
-                                .root_fields(&doc.executable)
-                                .filter(|field| field.name == "__typename")
-                                .map(|field| {
-                                    (
-                                        field.response_key().as_str().into(),
-                                        Value::String(root_type_name.clone()),
-                                    )
-                                })
-                                .collect(),
-                        );
-                        return Ok(QueryPlannerContent::Response {
-                            response: Box::new(graphql::Response::builder().data(data).build()),
-                        });
-                    } else {
-                        // fragments might use @include or @skip
-                    }
-                }
+        if doc.has_root_typename && !doc.has_schema_introspection && !doc.has_explicit_root_fields {
+            // Fast path for __typename alone
+            if doc
+                .operation
+                .selection_set
+                .selections
+                .iter()
+                .all(|sel| sel.as_field().is_some_and(|f| f.name == "__typename"))
```

Comment:
this could/should also use `operation.root_fields()`, i guess, but doesn't need to be addressed here

Reply:
`root_fields()` is used to compute `doc.has_root_typename` and implicitly descends into inline and named fragments regardless of any directive they might have. This specifically only looks at field, in order to not deal with potential `@skip` or `@include` that fragments might have.

… but now I realize that fields can also have `@skip` or `@include`, so I’ll add this restriction to the condition here.

Reply:
I’ll simplify this further in the next PR, having a fast path only for literally `{__typename}` and letting introspection code deal with any other combination of fragments, aliases, directives, etc.

Reply:
I figured that if we have a fast path for root typename, we might as well have a fast path for something like this?
```graphql
fragment tyname on Query { __typename }
{
  ... { ...tyname }
}
```

Though I wonder how useful a `__typename` fast path is in general with Rust introspection.